extendr-api/src/thread_safety.rs

```diff
@@ -1,69 +1,47 @@
 //! Provide limited protection for multithreaded access to the R API.
 
 use crate::*;
-use std::sync::atomic::{AtomicU32, Ordering};
+use std::cell::Cell;
+use std::sync::Mutex;
 
-static OWNER_THREAD: AtomicU32 = AtomicU32::new(0);
-static NEXT_THREAD_ID: AtomicU32 = AtomicU32::new(1);
+/// A global lock, that should represent the global lock on the R-API.
+/// It is not tied to an actual instance of R.
+static R_API_LOCK: Mutex<()> = Mutex::new(());
 
 thread_local! {
-    static THREAD_ID: u32 = NEXT_THREAD_ID.fetch_add(1, Ordering::SeqCst);
+    static THREAD_HAS_LOCK: Cell<bool> = Cell::new(false);
 }
```

Comment:
Why is this variable needed?

Reply:
It's due to nested calls of single threaded.
A thread acquires the lock, and performs calculations. That works fine.
If the same thread acquires the lock. And then it performs one calculation, but then the next thing is that it wants to acquire the lock again, then it is waiting in itself to release the lock to acquire it again. This variable tells it, if it already has the lock, I.e if it is already the thread that's allowed to access the r api at the moment.

This is the reentrant part.

Reply:
Ah, thanks, makes sense.

(What I felt a bit anxious was that, since this is not an atomic operation, if it can happen `has_lock` is `true` when checked here but is unlocked during `f()`. But, probably it won't happen as long as the things called sequentially.)

``` rust
    let _guard = if !has_lock {
        Some(R_API_LOCK.lock().unwrap())
    } else {
        None
    };
```

Reply:
Let me think about that.. I mean has lock is a property of the thread, so if won't be released by any other thread or anything, whilst f is running.

But it is incredibly difficult to reason about this.
Im still a bit unclear about the Cell part. I think that's the part that takes care of the "atomic" component...

Reply:
> Ah, thanks, makes sense.
> 
> (What I felt a bit anxious was that, since this is not an atomic operation, if it can happen `has_lock` is `true` when checked here but is unlocked during `f()`. But, probably it won't happen as long as the things called sequentially.)

The catch is that it's not even a "probably", it's a "can't".

```rs
    let has_lock = THREAD_HAS_LOCK.with(|x| x.get());

    // acquire R-API lock
    let _guard = if !has_lock {
        Some(R_API_LOCK.lock().unwrap())
    } else {
        None
    };

    // this thread now has the lock
    THREAD_HAS_LOCK.with(|x| x.set(true));

    let result = f();
```

The inherent-nature of a single thread of execution is that there are no code paths that can diverge into another `f()` (or anywhere at all) between the thread-local check, the mutex-lock, and the thread-local set. Other threads contending for mutex can't access the other thread-locals, only its own thread of execution's thread-local.

The worst-case-scenario is that a thread unwinds without releasing the lock. There's no UB, because the mutex will be poisoned for other threads (they wont be able to lock), and the original thread will always assume it already owns the lock.

Reply:
Yeah, you are right. Sorry if my wording made you feel uncomfortable.

Reply:
No offence taken! 

I think it is very difficult with this. I'm proposing an alternative mechanism, that somehow is failing CI,
and we have no way of truly figuring out if I'm write, or if there is another problem on extendr first.

Personally, I'm questioning everything.

I think this is pretty solid, so either (1) extendr is broken somewhere, and the `sleep` just masked that (2) conceptually, there is a missing piece.

In any case, reviews are welcome.